lib/extreme/persistent_subscription.ex

```diff
@@ -0,0 +1,325 @@
+  client (e.g. in `:dets` or PostgreSQL). Persistent Subscription is a
+  subscription strategy in which details about backpressure, buffer sizes, and
+  stream positions are all held by the EventStore (server).
+
+  In a persistent subscription, all communication is asynchronous. When an
+  event is received and processed, it must be acknowledged as processed by
+  `ack/3` in order to be considered by the server as processed. The server
+  stores knowledge of which events have been processed by means of checkpoints,
+  so listeners which use persistent subscriptions do not store stream positions
+  themselves.
+
+  The system of `ack/3`s and `nack/5`s allows listeners to handle events in
+  unconventional ways:
+
+  * concurrent processing: events may be handled by multiple processors at the
+    same time.
+  * out of order processing: events may be handled in any order.
+  * retry: events which are `nack/5`-ed with the `:retry` action, and events
+    which do not receive acknowledgement via `ack/3` are retried.
+  * message parking: if an event is not acknowledged and reaches its maximum
+    retry count, the message is parked in a parked messages queue. This
+    prevents head-of-line blocking typical of other subscription patterns.
+  * competing consumers: multiple consumers may process events without
+    collaboration or gossip between the consumers themselves.
+
+  Persistent subscriptions are started with
+  `c:Extreme.connect_to_persistent_subscription/4` expect a `cast` of each event
+  in the form of `{:on_event, event, correlation_id}`
+
+  A Persistent Subscription must exist before it can be connected to.
+  Persistent Subscriptions can be created by sending the
+  `Extreme.Messages.CreatePersistentSubscription` message via
+  `c:Extreme.execute/3`, by the HTTP API, or in the EventStore dashboard.
+
+  ## Example
+
+  ```elixir
+  defmodule MyPersistentListener do
+    use GenServer
+
+    alias Extreme.PersistentSubscription
+
+    def start_link(opts) do
+      GenServer.start_link(__MODULE__, opts)
+    end
+
+    def init(opts) do
+      {:ok, opts}
+    end
+
+    def subscribe(listener_proc), do: GenServer.cast(listener_proc, :subscribe)
+
+    def handle_cast(:subscribe, state) do
+      {:ok, subscription_pid} =
+        MyExtremeClientModule.connect_to_persistent_subscription(
+          self(),
+          opts.stream,
+          opts.group,
+          opts.allowed_in_flight_messages
+        )
+
+      {:noreply, Map.put(state, :subscription_pid, subscription_pid)}
+    end
+
+    def handle_cast({:on_event, event, correlation_id}, state) do
+      # .. do the real processing here ..
+
+      :ok = PersistentSubscription.ack(state.subscription_pid, event, correlation_id)
+
+      {:noreply, state}
+    end
+
+    def handle_cast(:unsubscribe, state) do
+      :ok = PersistentSubscription.unsubscribe(state.subscription_pid)
+
+      {:noreply, state}
+    end
```

Comment:
I think it'd be possible to implement unsubscription as a synchronous call (with `c:Extreme.unsubscribe/1`), and it might be desirable to do so for the sake of security around shutting down subscriptions. It'd be a refactor of moving most of the `_process_push`s from `Extreme.SharedSubscription` to this module, however...

actually as I type this I realize that that's not actually necessary and it's definitely better off as a synchronous call! (refactoring now...)